### ISSUE TITLE
File encryption

### ISSUE THREAD

**Issue by benridley** [projectsend/projectsend#311]
Hi guys I am new to this, but I was wondering if there is any way to encrypt the files stored in the directory used by Projectsend. 

Is there any advantage/weakness to this? What is the best way to protect sensitive files?

Thanks

**Comment by lenamtl** [comment]
To my opinion, you should never store any sensitive file on the web.
At the moment there are no file encryption.

What you can do for now is using a tool that encrypt file (manually), then upload the file encrypted.
Encrypting the file while uploading would be the best.
Something like this: PHP read the file contents, encrypt the contents, and save it to the file.

**Comment by tangy92** [comment]
This could work but what about the client receiving the document on the other side? If it's encrypted they're going to have an annoying time decrypting it unless it's seamless

**Comment by lenamtl** [comment]
If you encrypt them manually client/user will need a way to decrypt it.
If this does encrypt by the system it can be decrypted by the system on download.

But at the moment using PS there is no automatic way to do it.

**Comment by Mactory** [comment]
I am no security expert, but I am wondering about whether server side encryption is useful at all. What is the use-case where this makes things more secure?
I am guessing the scenario is, that somebody breaks into the server and tries to access the uploaded files. If they are encrypted, the attacker cannot read the contents of the files. However, if the server needs to be able to decrypt the files for downloading, it needs to store the key somewhere. If the attacker got into the server, he will probably also be able to retrieve the key from the filesystem or memory and hence gain access to the files.
Am I missing something? Is there a scenario where this type of encryption is feasible?

**Comment by tangy92** [comment]
I'm no expert either but I believe common practice is to store the keys on
another server. Like you said it's not perfect but it makes decrypting
files more inconvenient.

**Comment by shoulders** [comment]
on a limited basis you could encrypt the file locally and then email the encryption key to the admin of projectsend, then upload the file. project send will know it is encrypted (via a flag set during upload) and advise the admin on the other end they should have the key. hence no risk to the file being online. The key is not stored in projectsend avoiding hacking issues.

or use the standard public private key pair where there is a public key to encypt all data and requires the private key to decrypt it. this key would not be stored on project send making the transactions secure.

This feature would take quite a bit of programming I think.

**Comment by bpruett** [comment]
I agree with Mactory in terms of what happens in the case of a server breach. Give enough time on the server the code can be reverse engineered and the key found. 
That being said how useful the encryption is and how you go about it is really more a matter of your business requirements. 
With the fairly standard requirement of "encrypted in transit and at rest" there are a couple of options that could be done. 
Requiring SSL as the transport handles encryption in transit (a nice configuration option would be to require HTTPS and redirect any HTTP requests at the PHP level).
For encryption at rest, encrypted drives on the system handles those requirements (with user restrictions at the OS level). These would require no ProjectSend changes and meet the requirement.
I would like to see the option of file level encryption within the application, this could start with a system level key as a start and then progress to per user keys. This would protect the files (at least for a while) in the event of a breach.

**Comment by nv1t** [comment]
There is also another option: Encrypt/Decrypt it at Browser level. This should be feasable with smaller files. I don't have any idea if a larger file can be hold in memory for decryption, or if there is any "writing file" routing. 

The Key can be stored as # in the url, and has to be used by accessing the file.

This has been done already by various zerobin projects. (https://github.com/sebsauvage/ZeroBin)

**Comment by maltfield** [comment]
+1 for this feature! Honestly, I was very shocked to find that Project Send doesn't include symmetric file encryption at rest. Our org found the UX of FIleSend to be perfect for our needs (we create user accounts for our clients to send us super-sensitive documents containing lots of customer financial data), but then -- no encryption?

Ideally, this feature would:

1. Have an options in settings to *require encryption* of all files uploaded (ie: this wouldn't be an option to "choose" to encrypt at the time the client uploads the file)
1. Probably use AES256-GCM
1. Do the encryption in the browser so the server never knows the encryption key
1. Generate a long, random, symmetric encryption key for every client upload session and show it to the user just before they're given the option to upload files. The user should be prompted with a note that they should store the key safely and send it to the receiver of the files out-of-band because [a] Project Send will not be storing it and [b] neither the sender nor the receiver will be able to access the files without that key.

**Comment by ignacionelson** [comment]
@maltfield That's a great workflow!
It will break the ability to preview files, but I think it's OK if the note in the options page is clear enough.
I'll try to implement it this way, thanks!

**Comment by maltfield** [comment]
@ignacionelson also see how Lufi does this, which is almost exactly like I described.

 * https://demo.lufi.io/
 * https://framagit.org/fiat-tux/hat-softwares/lufi

It's written in perl, but the client-side encryption/decryption stuff is javascript using the Stanford Javascript Crypto Library (SJCL):

 * https://framagit.org/fiat-tux/hat-softwares/lufi/-/blob/master/themes/default/public/js/lufi-up.js
 * https://github.com/bitwiseshiftleft/sjcl

Also, I'd let the user specify the encryption key size in the Project Send config file.

**Comment by mdPlusPlus** [comment]
Has there been any progress on this in the meantime? I'm looking for a solution like ProjectSend, but e2e encryption is an absolute must.  
If this is not implemented yet, I need to evaluate other projects (like Peergos or myDrive).

**Comment by maltfield** [comment]
@mdPlusPlus fwiw, my org ended-up using Wire

 * https://wire.com/en/product/messaging/#file-sharing

We used their paid plan so our customers could communicate securely with us in their web browser (not require them to download any software). It wasn't the best solution, but it was definitely secure and checked all the compliance boxes.

**Comment by ignacionelson** [comment]
I had issues with memory limits while trying to add encryption.
I've found this https://medium.com/@antoine.lame/how-to-encrypt-files-with-php-f4adead297de which solved the problems and I have a local version running with this type of encryption enabled. This doesn't follow the proposed workflow, but I think that for the most use cases having to keep keys for every file is a bit complicated. (I do plan to add this as an option too so the user can select,  it's just a setting where the key used is the general one or a random, one time generated key)

**Comment by maltfield** [comment]
Sorry, but if encryption is done server-side (eg with PHP), then it defeats the whole purpose and you probably shouldn't bother writing it because nobody will want it.

It's very important to keep the server "Zero Knowledge" so its previously-sent files can't be compromised if the ProjectSend server is compromised..

 * https://en.wikipedia.org/wiki/Zero-knowledge_service

**Comment by ignacionelson** [comment]
I believe that "nobody will want it" is a bit much... Not all users want or require that much level of security. Hassle-free encryption (even if a breach means it's mostly useless) was a request by many of them so my plan is to have both of them available. This one was the easiest and quickest to implement, but it doesn't mean plans stop there.
There are big plans for PS in motion and I hope to cover as many of this compliance requirements soon.

**Comment by iincitr** [comment]
I agree with you Dear Nelson

4 Tem 2022 Pzt, saat 20:55 tarihinde Ignacio Nelson <
***@***.***> şunu yazdı:

**Comment by ignacionelson** [comment]
Started working on this one, both encryption types will be supported (php and js) for the user to select
![image](https://user-images.githubusercontent.com/6926562/179425628-7e858711-da47-4430-ad8d-1d14bcf02f92.png)

**Comment by maltfield** [comment]
Great news! I hope this wasn't forgotten

> Have an options in settings to require encryption of all files uploaded (ie: this wouldn't be an option to "choose" to encrypt at the time the client uploads the file)

So in this case, the "on" button should not be toggleable to "off" -- nor would a bypass be possible with some (malicious) user client changing POST or GET variables on that form submission.

If this "require client-side encryption" setting is enabled, then any attempts to upload a file without client-side encryption should fail with an error.

This is very important for orgs that have legal requirements to encrypt all files at-rest and in-transit (and face very large fines for violations).

**Comment by ignacionelson** [comment]
@maltfield Indeed, if encryption is required the toggle will not be switchable. It probably won't even be there, and instead have a text line about how the file will be forcefully encrypted

**Comment by ignacionelson** [comment]
A few more screens of how the UI is progressing
![image](https://user-images.githubusercontent.com/6926562/180328475-c88fc4e3-17cb-4d54-b7df-a14d12ab2b7d.png)

![image](https://user-images.githubusercontent.com/6926562/180328507-fb5a7764-f565-441e-a5c1-c9a8e8a8c25c.png)

**Comment by maltfield** [comment]
Wonderful to see progress on this, thank you!

It's interesting that it's a unique key per file. Is that a design choice?

Also, are you prompting the user to enter the key? No matter what helpful message you display, I'd bet >70% of the time a user will not record the key and they'll loose it the first time they use this software.

I'd make the key "disappear" on the screen, and then force the user to enter it again before proceeding with the upload. It's the same workflow used with setting up TOTP (eg Google Authenticator) and cryptocurrency wallets. If the key is stored by the client-only, you have to hold their hand and make sure they don't just click "next; next; next" without reading anything.

Maybe provide an option to disable this workflow somewhere in the application's settings, but I think the default should be to not let the user proceed with uploading an encrypted file if they don't type the symmetric key in correctly when it's not being displayed to them.

And if the symmetric key is distinct for every file on a bulk upload, maybe have a button for bulk `click to copy all` at the bottom that allows someone to get a line-separated list of `<key> <filename>\n<key> <filename>...` for all of the files above. That way after someone uploads 200 files they can do a 1-click copy and paste of everything they need into their password manager.

**Comment by ignacionelson** [comment]
@maltfield Thanks for the suggestions! Since multiple files can be uploaded at once, but they might end up being for different clients for example, I believe having a key for each file is a bit more secure, even if it takes more steps for the user.
I was thinking that the "Continue" button could be disabled until:
a- user clicks all "Copy key" buttons
b- clicks a single "I have copied all they keys" (maybe by selecting and copying all manually)

I'm also working in  the UI to let the user know that each file is being encrypted during the process.

A few issues due to my limited js:
- Encrypting large files is a problem since I'm not sure how to sync chunks between  js chunking and the upload process handled by plupload. *
I have successfully encrypted and uploaded several <10mb files at once, but each browser's limit might be an issue in this case. Also, this limits the functionality to documents and pictures only. Videos, audio, compressed zip files or other large ones will have to wait until I (or someone else) can solve this.
- Don't know how to (but I'm looking up documentation and examples) download a file in the background, then decrypt it and finally offer the regular download dialog to the user. I believe this is the only solution to keep the provided decryption key in the browser, and not send it in the request.

* Side note here: The current solution I've found is to catch each plupload file, encrypt it, remove the original, and add a new one (as a blob). Maybe the approach is wrong and I simply need to encrypt each existing plupload chunk as it is uploading, but haven't look into this yet.

I'm aiming to have this available as soon as I can, it's too important to not focus 100% on this.

**Comment by ignacionelson** [comment]
Update: seems like plupload chunks can be manipulated, so each chunk could possibly be encrypted before sending it, which solves the uploading  issues and most of the weird workarounds I made. I just need to find documentation (or check the code) to confirm this.

Edit: BeforeChunkUpload seems to be gone on plupload 3, so I will have to reimplement it somehow

**Comment by jwomackgsa** [comment]
Any updates on this feature?

**Comment by Tnology** [comment]
Anyone have an update for this?

**Comment by ignacionelson** [comment]
I haven't been able to implement on plupload. I've managed to encrypt server-side with a few caveats:
- encrypted file are larger than the originals
- the process takes time
- memory is a constraint, and this is the main issue. there are many users of PS that handle really large media files and those failed for this reason. maybe there's an alternative to php's encryption...? so only small files worked in my case.

**Comment by nv1t** [comment]
> I haven't been able to implement on plupload. I've managed to encrypt server-side with a few caveats:
> 
> * encrypted file are larger than the originals
> * the process takes time
> * memory is a constraint, and this is the main issue. there are many users of PS that handle really large media files and those failed for this reason. maybe there's an alternative to php's encryption...? so only small files worked in my case.

maybe it would be an option to implement [age](https://github.com/FiloSottile/age) on client side. there are javascript libraries for age. The only problem is handling the keys. Bu, it would be possible having public keys associated with the clients and they have private keys in their localstorage. The encryption would be done client side and the encrypted blob would be uploaded

**Comment by ignacionelson** [comment]
This issue is being closed as part of a bulk cleanup. ProjectSend has undergone significant changes since this was reported (including r1720 and r1994), and many older issues no longer apply to the current codebase.

If you are still experiencing this problem on the latest version, please open a new issue with updated details and we'll be happy to look into it. Thank you for your contribution!
